### ISSUE TITLE
Failing to create an instance

### ISSUE THREAD

**Issue by grzanka** [yaptide/deploy#5]
I ran the playbook [cloud_instance.yml](https://github.com/yaptide/deploy/blob/89b1af38c4743b84a586af0e5f2b7995ee375b1f/cloud_instance.yml) from https://github.com/yaptide/deploy/pull/4 on [3-add-roles-for-create-and-teardown-an-instance branch](https://github.com/yaptide/deploy/tree/89b1af38c4743b84a586af0e5f2b7995ee375b1f) and got following error.

```
(venv) grzanka@workspace:~/workspace/deploy$ ansible-playbook cloud_instance.yml
[WARNING]: Invalid characters were found in group names but not replaced, use -vvvv to see
details

PLAY [localhost] ****************************************************************************

TASK [openstack : Delete old key] ***********************************************************
ok: [localhost]

TASK [openstack : Create a keypair] *********************************************************
changed: [localhost]

TASK [openstack : Create the network] *******************************************************
changed: [localhost]

TASK [openstack : Create the subnet] ********************************************************
changed: [localhost]

TASK [openstack : Create the router] ********************************************************
changed: [localhost]

TASK [openstack : Create a new security group] **********************************************
changed: [localhost]

TASK [openstack : Create a new security group allowing any ICMP] ****************************
changed: [localhost]

TASK [openstack : Create a new security group allowing any SSH connection] ******************
changed: [localhost]

TASK [openstack : Launch an instance] *******************************************************
fatal: [localhost]: FAILED! => {"changed": false, "extra_data": {"data": null, "details": "None", "response": "None"}, "msg": "No port on server 2bcd0f73-abf2-4e75-a8ea-e56cbee37c77 was found matching your NAT destination network test_network. Please  check your config"}

PLAY RECAP **********************************************************************************
localhost                  : ok=8    changed=7    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
```

Despite the error the instance with id `2bcd0f73-abf2-4e75-a8ea-e56cbee37c77` was created.
Two facts:
 - the instance didn't get a public IP address
 - two regular network interfaces were attached

![Screenshot 2022-10-18 at 11 40 20](https://user-images.githubusercontent.com/7374733/196395437-0cab1d3f-352e-4eef-b538-d4eb2752f9df.png)

Log file of the machine:
```
[   22.193481] cloud-init[559]: Cloud-init v. 20.3-2-g371b392c-0ubuntu1~20.04.1 running 'init' at Tue, 18 Oct 2022 09:35:48 +0000. Up 22.00 seconds.
[   22.198892] cloud-init[559]: ci-info: +++++++++++++++++++++++++++++++++++++++Net device info+++++++++++++++++++++++++++++++++++++++
[   22.203557] cloud-init[559]: ci-info: +--------+------+------------------------------+---------------+--------+-------------------+
[   22.208262] cloud-init[559]: ci-info: | Device |  Up  |           Address            |      Mask     | Scope  |     Hw-Address    |
[   22.212838] cloud-init[559]: ci-info: +--------+------+------------------------------+---------------+--------+-------------------+
[   22.217424] cloud-init[559]: ci-info: |  ens3  | True |          10.0.0.11           | 255.255.255.0 | global | fa:16:3e:bc:3c:ed |
[   22.222126] cloud-init[559]: ci-info: |  ens3  | True | fe80::f816:3eff:febc:3ced/64 |       .       |  link  | fa:16:3e:bc:3c:ed |
[   22.227022] cloud-init[559]: ci-info: |  ens4  | True |          10.0.0.14           | 255.255.255.0 | global | fa:16:3e:7b:5b:b6 |
[   22.231840] cloud-init[559]: ci-info: |  ens4  | True | fe80::f816:3eff:fe7b:5bb6/64 |       .       |  link  | fa:16:3e:7b:5b:b6 |
[   22.236698] cloud-init[559]: ci-info: |   lo   | True |          127.0.0.1           |   255.0.0.0   |  host  |         .         |
[   22.242350] cloud-init[559]: ci-info: |   lo   | True |           ::1/128            |       .       |  host  |         .         |
[   22.246607] cloud-init[559]: ci-info: +--------+------+------------------------------+---------------+--------+-------------------+
[   22.250731] cloud-init[559]: ci-info: ++++++++++++++++++++++++++++++Route IPv4 info+++++++++++++++++++++++++++++++
[   22.254507] cloud-init[559]: ci-info: +-------+-----------------+----------+-----------------+-----------+-------+
[   22.258106] cloud-init[559]: ci-info: | Route |   Destination   | Gateway  |     Genmask     | Interface | Flags |
[   22.261640] cloud-init[559]: ci-info: +-------+-----------------+----------+-----------------+-----------+-------+
[   22.265220] cloud-init[559]: ci-info: |   0   |     0.0.0.0     | 10.0.0.1 |     0.0.0.0     |    ens4   |   UG  |
[   22.268732] cloud-init[559]: ci-info: |   1   |     0.0.0.0     | 10.0.0.1 |     0.0.0.0     |    ens3   |   UG  |
[   22.272280] cloud-init[559]: ci-info: |   2   |     10.0.0.0    | 0.0.0.0  |  255.255.255.0  |    ens4   |   U   |
[   22.275799] cloud-init[559]: ci-info: |   3   |     10.0.0.0    | 0.0.0.0  |  255.255.255.0  |    ens3   |   U   |
[   22.279357] cloud-init[559]: ci-info: |   4   | 169.254.169.254 | 10.0.0.1 | 255.255.255.255 |    ens4   |  UGH  |
[   22.282962] cloud-init[559]: ci-info: |   5   | 169.254.169.254 | 10.0.0.1 | 255.255.255.255 |    ens3   |  UGH  |
[   22.286475] cloud-init[559]: ci-info: +-------+-----------------+----------+-----------------+-----------+-------+
[   22.289999] cloud-init[559]: ci-info: +++++++++++++++++++Route IPv6 info+++++++++++++++++++
[   22.292839] cloud-init[559]: ci-info: +-------+-------------+---------+-----------+-------+
[   22.295773] cloud-init[559]: ci-info: | Route | Destination | Gateway | Interface | Flags |
[   22.298725] cloud-init[559]: ci-info: +-------+-------------+---------+-----------+-------+
[   22.301654] cloud-init[559]: ci-info: |   1   |  fe80::/64  |    ::   |    ens4   |   U   |
[   22.304717] cloud-init[559]: ci-info: |   2   |  fe80::/64  |    ::   |    ens3   |   U   |
[   22.307763] cloud-init[559]: ci-info: |   4   |    local    |    ::   |    ens4   |   U   |
[   22.310763] cloud-init[559]: ci-info: |   5   |    local    |    ::   |    ens3   |   U   |
[   22.313640] cloud-init[559]: ci-info: |   6   |   ff00::/8  |    ::   |    ens4   |   U   |
[   22.316519] cloud-init[559]: ci-info: |   7   |   ff00::/8  |    ::   |    ens3   |   U   |
[   22.319445] cloud-init[559]: ci-info: +-------+-------------+---------+-----------+-------+
```

**Comment by grzanka** [comment]
I've changed the approach to add external IP in a separate step, as shown here:
https://github.com/yaptide/deploy/pull/4/commits/375dad8fcf52971db8d3baa0fda4855c8b40efc0#diff-0145daefdae8ff027351c2b158135bb3df7eafb1e4f1063c4bb8e5530f05ca6eR66

I've cleaned the instance (running `ansible-playbook cloud_instance.yml -e mode=clean`) and rerun the playbook again.
This time everything seemed OK:

```
(venv) grzanka@workspace:~/workspace/deploy$ ansible-playbook cloud_instance.yml
[WARNING]: Invalid characters were found in group names but not replaced, use -vvvv to see details

PLAY [localhost] ************************************************************************************************

TASK [openstack : Delete old key] *******************************************************************************
changed: [localhost]

TASK [openstack : Create a keypair] *****************************************************************************
changed: [localhost]

TASK [openstack : Create the network] ***************************************************************************
changed: [localhost]

TASK [openstack : Create the subnet] ****************************************************************************
changed: [localhost]

TASK [openstack : Create the router] ****************************************************************************
changed: [localhost]

TASK [openstack : Create a new security group] ******************************************************************
changed: [localhost]

TASK [openstack : Create a new security group allowing any ICMP] ************************************************
changed: [localhost]

TASK [openstack : Create a new security group allowing any SSH connection] **************************************
changed: [localhost]

TASK [openstack : Launch an instance] ***************************************************************************
changed: [localhost]

TASK [openstack : openstack.cloud.floating_ip] ******************************************************************
changed: [localhost]

TASK [openstack : Show IP] **************************************************************************************
ok: [localhost] => {
    "msg": "149.156.182.250"
}

TASK [openstack : Scan for ssh public key] **********************************************************************
FAILED - RETRYING: [localhost]: Scan for ssh public key (10 retries left).
FAILED - RETRYING: [localhost]: Scan for ssh public key (9 retries left).
FAILED - RETRYING: [localhost]: Scan for ssh public key (8 retries left).
FAILED - RETRYING: [localhost]: Scan for ssh public key (7 retries left).
ok: [localhost]

TASK [openstack : Add/update the public key(s) in the known_hosts file] *****************************************
changed: [localhost]

TASK [openstack : Delete an instance] ***************************************************************************
skipping: [localhost]

TASK [openstack : Delete security group] ************************************************************************
skipping: [localhost]

TASK [openstack : Delete the router] ****************************************************************************
skipping: [localhost]

TASK [openstack : Delete the network] ***************************************************************************
skipping: [localhost]

PLAY RECAP ******************************************************************************************************
localhost                  : ok=13   changed=11   unreachable=0    failed=0    skipped=4    rescued=0    ignored=0   

(venv) grzanka@workspace:~/workspace/deploy$ ssh ubuntu@149.156.182.250
Welcome to Ubuntu 20.04.1 LTS (GNU/Linux 5.4.0-51-generic x86_64)

 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/advantage

  System information as of Tue Oct 18 10:05:02 UTC 2022

  System load:  1.48              Processes:             121
  Usage of /:   5.1% of 24.06GB   Users logged in:       0
  Memory usage: 5%                IPv4 address for ens3: 10.0.0.8
  Swap usage:   0%

1 update can be installed immediately.
0 of these updates are security updates.
To see these additional updates run: apt list --upgradable


The list of available updates is more than a week old.
To check for new updates run: sudo apt update


The programs included with the Ubuntu system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Ubuntu comes with ABSOLUTELY NO WARRANTY, to the extent permitted by
applicable law.

To run a command as administrator (user "root"), use "sudo <command>".
See "man sudo_root" for details.

ubuntu@yaptide:~$ 
```

**Comment by grzanka** [comment]
After several upgrades, see https://github.com/yaptide/deploy/pull/4.

The playbook fails:
```
(...)
TASK [openstack : Launch an instance] **************************************************************************************************************************************************************************
changed: [localhost]

TASK [openstack : Assign a floating IP to the instance] ********************************************************************************************************************************************************
changed: [localhost]

TASK [openstack : Show IP] *************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "149.156.182.250"
}

TASK [openstack : Scan for ssh public key] *********************************************************************************************************************************************************************
FAILED - RETRYING: [localhost]: Scan for ssh public key (10 retries left).
FAILED - RETRYING: [localhost]: Scan for ssh public key (9 retries left).
FAILED - RETRYING: [localhost]: Scan for ssh public key (8 retries left).
FAILED - RETRYING: [localhost]: Scan for ssh public key (7 retries left).
FAILED - RETRYING: [localhost]: Scan for ssh public key (6 retries left).
FAILED - RETRYING: [localhost]: Scan for ssh public key (5 retries left).
FAILED - RETRYING: [localhost]: Scan for ssh public key (4 retries left).
FAILED - RETRYING: [localhost]: Scan for ssh public key (3 retries left).
FAILED - RETRYING: [localhost]: Scan for ssh public key (2 retries left).
FAILED - RETRYING: [localhost]: Scan for ssh public key (1 retries left).
fatal: [localhost]: FAILED! => {"attempts": 10, "changed": true, "cmd": ["ssh-keyscan", "-H", "149.156.182.250"], "delta": "0:00:00.323264", "end": "2022-10-18 12:45:25.641975", "msg": "non-zero return code", "rc": 1, "start": "2022-10-18 12:45:25.318711", "stderr": "write (149.156.182.250): No route to host\r\nwrite (149.156.182.250): No route to host\r\nwrite (149.156.182.250): No route to host\r\nwrite (149.156.182.250): No route to host\r\nwrite (149.156.182.250): No route to host", "stderr_lines": ["write (149.156.182.250): No route to host", "write (149.156.182.250): No route to host", "write (149.156.182.250): No route to host", "write (149.156.182.250): No route to host", "write (149.156.182.250): No route to host"], "stdout": "", "stdout_lines": []}

PLAY RECAP *****************************************************************************************************************************************************************************************************
localhost                  : ok=11   changed=10   unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
```

The instance was created again with two network interfaces:
![image](https://user-images.githubusercontent.com/7374733/196435610-430e69af-54ca-42eb-b7df-67d46a047096.png)


I've got again an error in the log file:
```
         Starting [0;1;39mLoad AppArmor pro���managed internally by snapd[0m...
         Starting [0;1;39mInitial cloud-init job (pre-networking)[0m...
[[0;32m  OK  [0m] Finished [0;1;39mLoad AppArmor pro���s managed internally by snapd[0m.
[   19.713618] cloud-init[496]: Cloud-init v. 20.3-2-g371b392c-0ubuntu1~20.04.1 running 'init-local' at Tue, 18 Oct 2022 12:43:31 +0000. Up 16.42 seconds.
[   19.718840] cloud-init[496]: 2022-10-18 12:43:35,135 - util.py[WARNING]: failed stage init-local
[   19.733132] cloud-init[496]: failed run of stage init-local
[   19.741504] cloud-init[496]: ------------------------------------------------------------
[   19.743896] cloud-init[496]: Traceback (most recent call last):
[   19.745772] cloud-init[496]:   File "/usr/lib/python3/dist-packages/cloudinit/cmd/main.py", line 653, in status_wrapper
[   19.748898] cloud-init[496]:     ret = functor(name, args)
[   19.750633] cloud-init[496]:   File "/usr/lib/python3/dist-packages/cloudinit/cmd/main.py", line 362, in main_init
[   19.753692] cloud-init[496]:     init.apply_network_config(bring_up=bool(mode != sources.DSMODE_LOCAL))
[   19.756632] cloud-init[496]:   File "/usr/lib/python3/dist-packages/cloudinit/stages.py", line 678, in apply_network_config
[   19.759881] cloud-init[496]:     netcfg, src = self._find_networking_config()
[   19.761995] cloud-init[496]:   File "/usr/lib/python3/dist-packages/cloudinit/stages.py", line 643, in _find_networking_config
[   19.765258] cloud-init[496]:     if self.datasource and hasattr(self.datasource, 'network_config'):
[   19.767797] cloud-init[496]:   File "/usr/lib/python3/dist-packages/cloudinit/sources/DataSourceOpenStack.py", line 117, in network_config
[   19.771768] cloud-init[496]:     self._network_config = openstack.convert_net_json(
[   19.774089] cloud-init[496]:   File "/usr/lib/python3/dist-packages/cloudinit/sources/helpers/openstack.py", line 687, in convert_net_json
[   19.777577] cloud-init[496]:     raise ValueError("Unable to find a system nic for %s" % d)
[   19.780003] cloud-init[496]: ValueError: Unable to find a system nic for {'type': 'physical', 'mtu': 8950, 'subnets': [{'type': 'dhcp4'}], 'mac_address': 'fa:16:3e:07:c5:ec'}
[   19.784466] cloud-init[496]: ------------------------------------------------------------
[[0;1;31mFAILED[0m] Failed to start [0;1;39mInitial cloud-init job (pre-networking)[0m.
See 'systemctl status cloud-init-local.service' for details.
[[0;32m  OK  [0m] Reached target [0;1;39mNetwork (Pre)[0m.
         Starting [0;1;39mNetwork Service[0m...
[[0;32m  OK  [0m] Started [0;1;39mNetwork Service[0m.
         Starting [0;1;39mWait for Network to be Configured[0m...
         Starting [0;1;39mNetwork Name Resolution[0m...
[[0;32m  OK  [0m] Finished [0;1;39mWait for Network to be Configured[0m.
         Starting [0;1;39mInitial cloud-ini��� (metadata service crawler)[0m...
[[0;32m  OK  [0m] Started [0;1;39mNetwork Name Resolution[0m.
[[0;32m  OK  [0m] Reached target [0;1;39mNetwork[0m.
[[0;32m  OK  [0m] Reached target [0;1;39mHost and Network Name Lookups[0m.
[   20.568181] cloud-init[526]: Cloud-init v. 20.3-2-g371b392c-0ubuntu1~20.04.1 running 'init' at Tue, 18 Oct 2022 12:43:35 +0000. Up 20.38 seconds.
[   20.572960] cloud-init[526]: ci-info: +++++++++++++++++++++++++++Net device info++++++++++++++++++++++++++++
[   20.576467] cloud-init[526]: ci-info: +--------+-------+-----------+-----------+-------+-------------------+
[   20.580079] cloud-init[526]: ci-info: | Device |   Up  |  Address  |    Mask   | Scope |     Hw-Address    |
[   20.583462] cloud-init[526]: ci-info: +--------+-------+-----------+-----------+-------+-------------------+
[   20.586813] cloud-init[526]: ci-info: |  ens3  | False |     .     |     .     |   .   | fa:16:3e:fc:2c:76 |
[   20.590190] cloud-init[526]: ci-info: |   lo   |  True | 127.0.0.1 | 255.0.0.0 |  host |         .         |
[   20.593631] cloud-init[526]: ci-info: |   lo   |  True |  ::1/128  |     .     |  host |         .         |
[   20.597022] cloud-init[526]: ci-info: +--------+-------+-----------+-----------+-------+-------------------+
[   20.600385] cloud-init[526]: ci-info: +++++++++++++++++++Route IPv6 info+++++++++++++++++++
[   20.603339] cloud-init[526]: ci-info: +-------+-------------+---------+-----------+-------+
[   20.606304] cloud-init[526]: ci-info: | Route | Destination | Gateway | Interface | Flags |
[   20.609183] cloud-init[526]: ci-info: +-------+-------------+---------+-----------+-------+
[   20.612185] cloud-init[526]: ci-info: +-------+-------------+---------+-----------+-------+[[0;1;31mFAILED[0m] Failed to start [0;1;39mInitial cl���ob (metadata service crawler)[0m.
See 'systemctl status cloud-init.service' for details.
[[0;32m  OK  [0m] Reached target [0;1;39mCloud-config availability[0m.
[   20.623007] cloud-init[526]: 2022-10-18 12:43:35,990 - util.py[WARNING]: failed stage init
```

**Comment by grzanka** [comment]
Fixed in #4
